[SUPERSEDED] Unit companion is compileTimeOnly

som-snytt opened the pull request:
Annotations get a dispensation for the ref check.

The `Group` mechanism for specialized only uses the type args, so avoid using `Unit` to define a value.

Forward ports https://github.com/scala/scala/pull/6490, which didn't cope with `Group` definitions.

som-snytt committed:
Setup Unit companion as CompileTimeOnly

Annotations get a dispensation for the ref check.

The `Group` mechanism for specialized is moved
to the type level.

lihaoyi commented:
If you don't want people to access the unit companion object `Unit`, why not just rename it from `Unit` into something like `UnitLolz`? Seems easier than trying to do funny things with types inside the compiler internals

som-snytt commented:
@lihaoyi I don't disagree with you, see https://twitter.com/dwijnand/status/1073149475936485376 which says apparently anyone can confuse it.

I'm just trying to help the folks on current version of Scala, using available mechanisms.

hmemcpy commented:
Heh, even @odersky confused them on stage at ScalaExchange :)

som-snytt committed:
Unleash Unit restriction

Add some convenience groups to Specializable and
avoid usages of Unit companion in library so that
Unit companion can be compileTimeOnly without
compiler assistance.

som-snytt commented:
Of course, this makes it harder to test methods on Unit.

som-snytt committed:
Regenerate

som-snytt commented:
Probably this PR had seniority over recent efforts to leave fingerprints on `Function0`.
